README.md
# PHP DSN Parser

[![Release](https://img.shields.io/github/v/release/ixnode/php-dsn-parser)](https://github.com/ixnode/php-dsn-parser/releases)
[![](https://img.shields.io/github/release-date/ixnode/php-dsn-parser)](https://github.com/ixnode/php-dsn-parser/releases)
![](https://img.shields.io/github/repo-size/ixnode/php-dsn-parser.svg)
[![PHP](https://img.shields.io/badge/PHP-^8.2-777bb3.svg?logo=php&logoColor=white&labelColor=555555&style=flat)](https://www.php.net/supported-versions.php)
[![PHPStan](https://img.shields.io/badge/PHPStan-Level%20Max-777bb3.svg?style=flat)](https://phpstan.org/user-guide/rule-levels)
[![PHPUnit](https://img.shields.io/badge/PHPUnit-Unit%20Tests-6b9bd2.svg?style=flat)](https://phpunit.de)
[![PHPCS](https://img.shields.io/badge/PHPCS-PSR12-416d4e.svg?style=flat)](https://www.php-fig.org/psr/psr-12/)
[![PHPMD](https://img.shields.io/badge/PHPMD-ALL-364a83.svg?style=flat)](https://github.com/phpmd/phpmd)
[![Rector - Instant Upgrades and Automated Refactoring](https://img.shields.io/badge/Rector-PHP%208.2-73a165.svg?style=flat)](https://github.com/rectorphp/rector)
[![LICENSE](https://img.shields.io/github/license/ixnode/php-api-version-bundle)](https://github.com/ixnode/php-api-version-bundle/blob/master/LICENSE)

> This library helps to parse DSN\URI strings.

## 1. Usage

```php
use Ixnode\PhpDsnParser\DsnParser;
```

```php
$dsnParser = new DsnParser('smtp://suserweb:S22jD7Po%.,/zu34k@mail.domain.tld:25?verify_peer=0');

print_r($dsnParser->getParsed());
// (array) [
//     'protocol' => 'smtp',
//     'user' => 'suserweb',
//     'password' => 'S22jD7Po%.,/zu34k',
//     'host' => 'mail.domain.tld',
//     'port' => 25,
//     'options' => 'verify_peer=0',
// ]
```

## 2. Installation

```bash
composer require ixnode/php-dsn-parser
```

```bash
vendor/bin/php-dsn-parser -V
```

```bash
php-dsn-parser 0.1.0 (03-07-2023 01:17:26) - Björn Hempel <bjoern@hempel.li>
```

## 3. Library development

```bash
git clone git@github.com:ixnode/php-dsn-parser.git && cd php-dsn-parser
```

```bash
composer install
```

```bash
composer test
```

## 4. License

This library is licensed under the MIT License - see the [LICENSE](/LICENSE) file for details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Releases

### [0.1.1] - 2023-09-16

* Update README.md

### [0.1.0] - 2023-09-16

* Initial release with first DSN parser
* Add src
* Add tests
  * PHP Coding Standards Fixer
  * PHPMND - PHP Magic Number Detector
  * PHPStan - PHP Static Analysis Tool
  * PHPUnit - The PHP Testing Framework
  * Rector - Instant Upgrades and Automated Refactoring
* Add README.md
* Add LICENSE.md

## Add new version

```bash
# Checkout master branch
$ git checkout main && git pull

# Check current version
$ vendor/bin/version-manager --current

# Increase patch version
$ vendor/bin/version-manager --patch

# Change changelog
$ vi CHANGELOG.md

# Push new version
$ git add CHANGELOG.md VERSION && git commit -m "Add version $(cat VERSION)" && git push

# Tag and push new version
$ git tag -a "$(cat VERSION)" -m "Version $(cat VERSION)" && git push origin "$(cat VERSION)"
```


tests/Unit/DsnParserTest.php
<?php

/*
 * This file is part of the ixnode/php-dsn-parser project.
 *
 * (c) Björn Hempel <https://www.hempel.li/>
 *
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 */

declare(strict_types=1);

namespace Ixnode\PhpDsnParser\Tests\Unit;

use Ixnode\PhpDsnParser\DsnParser;
use PHPUnit\Framework\TestCase;

/**
 * Class DsnParserTest
 *
 * @author Björn Hempel <bjoern@hempel.li>
 * @version 0.1.0 (2023-09-16)
 * @since 0.1.0 (2023-09-16) First version.
 * @link DsnParser
 */
final class DsnParserTest extends TestCase
{
    /**
     * Test wrapper.
     *
     * @dataProvider dataProvider
     *
     * @test
     * @testdox $number) Test DsnParser:getParsed
     * @param int $number
     * @param string $dsn
     * @param array<string, string|null> $expected
     */
    public function wrapper(int $number, string $dsn, array $expected): void
    {
        /* Arrange */

        /* Act */
        $dsnParser = new DsnParser($dsn);

        /* Assert */
        $this->assertIsNumeric($number); // To avoid phpmd warning.
        $this->assertSame($expected, $dsnParser->getParsed());
    }

    /**
     * Data provider.
     *
     * @return array<int, mixed>
     */
    public function dataProvider(): array
    {
        $number = 0;

        return [
            [++$number, 'smtp://host:1025', [
                'protocol' => 'smtp',
                'user' => null,
                'password' => null,
                'host' => 'host',
                'port' => 1025,
                'options' => null,
            ], ],
            [++$number, 'smtp://abc:xyz@test:123', [
                'protocol' => 'smtp',
                'user' => 'abc',
                'password' => 'xyz',
                'host' => 'test',
                'port' => 123,
                'options' => null,
            ], ],
            [++$number, 'smtp://suserweb:S22jD7Po%.,/zu34j@mail.domain.tld:25', [
                'protocol' => 'smtp',
                'user' => 'suserweb',
                'password' => 'S22jD7Po%.,/zu34j',
                'host' => 'mail.domain.tld',
                'port' => 25,
                'options' => null,
            ], ],
            [++$number, 'smtp://suserweb:S22jD7Po%.,/zu34j@mail.domain.tld:25?verify_peer=0', [
                'protocol' => 'smtp',
                'user' => 'suserweb',
                'password' => 'S22jD7Po%.,/zu34j',
                'host' => 'mail.domain.tld',
                'port' => 25,
                'options' => 'verify_peer=0',
            ], ],
        ];
    }
}


rector.php
<?php

/*
 * This file is part of the ixnode/php-dsn-parser project.
 *
 * (c) Björn Hempel <https://www.hempel.li/>
 *
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 */

declare(strict_types=1);

use Rector\CodeQuality\Rector\Class_\InlineConstructorDefaultToPropertyRector;
use Rector\Config\RectorConfig;
use Rector\Doctrine\Set\DoctrineSetList;
use Rector\Php80\Rector\FunctionLike\MixedTypeRector;
use Rector\Php80\Rector\FunctionLike\UnionTypesRector;
use Rector\Set\ValueObject\LevelSetList;
use Rector\Symfony\Set\SensiolabsSetList;
use Rector\Symfony\Set\SymfonySetList;

return static function (RectorConfig $rectorConfig): void {
    $rectorConfig->paths([
        __DIR__ . '/src',
        __DIR__ . '/tests',
    ]);

    $rectorConfig->rule(InlineConstructorDefaultToPropertyRector::class);

    $rectorConfig->sets([
        LevelSetList::UP_TO_PHP_82,
        DoctrineSetList::ANNOTATIONS_TO_ATTRIBUTES,
        SymfonySetList::ANNOTATIONS_TO_ATTRIBUTES,
        SensiolabsSetList::FRAMEWORK_EXTRA_61,
    ]);

    /* Ignore additional type hints (param, return) within doc blocks */
    $rectorConfig->skip([
        MixedTypeRector::class,
        UnionTypesRector::class,
    ]);
};


src/DsnParser.php
<?php

/*
 * This file is part of the ixnode/php-dsn-parser project.
 *
 * (c) Björn Hempel <https://www.hempel.li/>
 *
 * For the full copyright and license information, please view the LICENSE.md
 * file that was distributed with this source code.
 */

declare(strict_types=1);

namespace Ixnode\PhpDsnParser;

use Ixnode\PhpDsnParser\Tests\Unit\DsnParserTest;

/**
 * Class DsnParser
 *
 * @author Björn Hempel <bjoern@hempel.li>
 * @version 0.1.0 (2023-09-16)
 * @since 0.1.0 (2023-09-16) First version.
 * @link DsnParserTest
 */
class DsnParser
{
    private const REGEXP_DSN = '~^([a-z]+)://(?:(?=.*@)([^:]+):([^@]+)@)?([^:]+):([0-9]+)(?:(?:\?)?(.+|))~';

    private const KEY_PROTOCOL = 'protocol';

    private const KEY_USER = 'user';

    private const KEY_PASSWORD = 'password';

    private const KEY_HOST = 'host';

    private const KEY_PORT = 'port';

    private const KEY_OPTIONS = 'options';

    /** @var array{protocol: string|null, user: string|null, password: string|null, host: string|null, port: int|null, options: string|null}|false $parsed */
    private readonly array|false $parsed;

    /**
     * @param string $dsn
     */
    public function __construct(protected string $dsn)
    {
        $this->parsed = $this->doParse($dsn);
    }

    /**
     * Parses the given dsn string.
     *
     * @param string $dsn
     * @return array{protocol: string|null, user: string|null, password: string|null, host: string|null, port: int|null, options: string|null}|false
     */
    private function doParse(string $dsn): array|false
    {
        $parts = [];

        if (!preg_match(self::REGEXP_DSN, $dsn, $parts)) {
            return false;
        }

        return [
            self::KEY_PROTOCOL => $this->getStringNull($parts, 1),
            self::KEY_USER => $this->getStringNull($parts, 2),
            self::KEY_PASSWORD => $this->getStringNull($parts, 3),
            self::KEY_HOST => $this->getStringNull($parts, 4),
            self::KEY_PORT => $this->getIntNull($parts, 5),
            self::KEY_OPTIONS => $this->getStringNull($parts, 6),
        ];
    }

    /**
     * Returns a part from given index (as string representation).
     *
     * @param array<int, string> $parts
     * @param int $index
     * @return string|null
     */
    private function getStringNull(array $parts, int $index): ?string
    {
        if (!array_key_exists($index, $parts)) {
            return null;
        }

        $part = $parts[$index];

        if (empty($part)) {
            return null;
        }

        return strval($part);
    }

    /**
     * Returns a part from given index (as string representation).
     *
     * @param array<int, string> $parts
     * @param int $index
     * @return int|null
     */
    private function getIntNull(array $parts, int $index): ?int
    {
        if (!array_key_exists($index, $parts)) {
            return null;
        }

        $part = $parts[$index];

        if (empty($part)) {
            return null;
        }

        return intval($part);
    }

    /**
     * Returns only one part of parsed data.
     *
     * @param string $key
     * @return string|int|null
     */
    private function getX(string $key): string|int|null
    {
        $parsed = $this->getParsed();

        if ($parsed === false) {
            return null;
        }

        if (!array_key_exists($key, $parsed)) {
            return null;
        }

        return $parsed[$key];
    }

    /**
     * Returns the parsed dsn representation.
     *
     * @return array{protocol: string|null, user: string|null, password: string|null, host: string|null, port: int|null, options: string|null}|false
     */
    public function getParsed(): array|false
    {
        return $this->parsed;
    }

    /**
     * Returns the protocol.
     *
     * @return string|null
     */
    public function getProtocol(): string|null
    {
        $value = $this->getX(self::KEY_PROTOCOL);

        if (is_null($value)) {
            return null;
        }

        return strval($value);
    }

    /**
     * Returns the user.
     *
     * @return string|null
     */
    public function getUser(): string|null
    {
        $value = $this->getX(self::KEY_USER);

        if (is_null($value)) {
            return null;
        }

        return strval($value);
    }

    /**
     * Returns the password.
     *
     * @return string|null
     */
    public function getPassword(): string|null
    {
        $value = $this->getX(self::KEY_PASSWORD);

        if (is_null($value)) {
            return null;
        }

        return strval($value);
    }

    /**
     * Returns the host.
     *
     * @return string|null
     */
    public function getHost(): string|null
    {
        $value = $this->getX(self::KEY_HOST);

        if (is_null($value)) {
            return null;
        }

        return strval($value);
    }

    /**
     * Returns the port.
     *
     * @return int|null
     */
    public function getPort(): int|null
    {
        $value = $this->getX(self::KEY_PORT);

        if (is_null($value)) {
            return null;
        }

        return intval($value);
    }

    /**
     * Returns the options as string.
     *
     * @return string|null
     */
    public function getOptions(): string|null
    {
        $value = $this->getX(self::KEY_OPTIONS);

        if (is_null($value)) {
            return null;
        }

        return strval($value);
    }
}
